### ISSUE TITLE
variable.data returns a variable, but doesn't share version counters

### ISSUE THREAD

**Issue by apaszke** [pytorch/pytorch#5396]
It just unwraps and re-wraps the `.data()` in C++, but does nothing about the version counter.

**Comment by apaszke** [comment]
cc: @colesbury

**Comment by colesbury** [comment]
This was mainly to keep backwards compatibility, but I'm not sure what behavior we want long term. I think `.data` with version counters would be identical `.detach()`.

**Comment by ezyang** [comment]
Recording conversation with @colesbury 

One (undocumented) perk of having `variable.data` not share version counters is that it is sometimes useful to explicitly unshare version counters. For example, suppose you have a single buffer, and you make disjoint views into the buffer. From the perspective of the user, mutating one view shouldn't affect others, and you don't want the version counters to line up. This is better than a different escape hatch of a guard which disables version counter updates, because that guard must be done at mutation site, whereas unsharing of version counters can be done at creation site and then forgotten about.

One reason to "fix" this behavior is if you want to store the version counter directly on storage; in that case, there is no way to unshare the counters. But it is only a very modest improvement, so we'll probably keep it possible to unshare counters, even if we do fix `variable.data` (but there probably isn't much reason to fix `variable.data` then, either.

cc @yf225

**Comment by albanD** [comment]
We are in the process of removing .data https://github.com/pytorch/pytorch/issues/30987 so this particular issue won't be solved.
